stage1/init/path.go

```diff
@@ -31,56 +32,59 @@ const (
 	socketsWantsDir = unitsDir + "/sockets.target.wants"
 )
 
-// ServiceUnitName returns a systemd service unit name for the given imageID
-func ServiceUnitName(imageID types.Hash) string {
-	return types.ShortHash(imageID.String()) + ".service"
+// ServiceUnitName returns a systemd service unit name based on the position of the
+// app in the pod manifest.
+func ServiceUnitName(index int) string {
+	return fmt.Sprintf("app-%d.service", index)
 }
 
-// ServiceUnitPath returns the path to the systemd service file for the given
-// imageID
-func ServiceUnitPath(root string, imageID types.Hash) string {
-	return filepath.Join(common.Stage1RootfsPath(root), unitsDir, ServiceUnitName(imageID))
+// ServiceUnitPath returns the path to the systemd service file based on the position
+// of the app in the pod manifest.
+func ServiceUnitPath(root string, index int) string {
+	return filepath.Join(common.Stage1RootfsPath(root), unitsDir, ServiceUnitName(index))
 }
 
-// RelEnvFilePath returns the path to the environment file for the given imageID
-// relative to the pod's root
-func RelEnvFilePath(imageID types.Hash) string {
-	return filepath.Join(envDir, types.ShortHash(imageID.String()))
+// RelEnvFilePath returns the path to the environment file relative to the pod's root.
+func RelEnvFilePath(index int) string {
+	return filepath.Join(envDir, strconv.Itoa(index))
 }
 
-// EnvFilePath returns the path to the environment file for the given imageID
-func EnvFilePath(root string, imageID types.Hash) string {
-	return filepath.Join(common.Stage1RootfsPath(root), RelEnvFilePath(imageID))
+// EnvFilePath returns the path to the environment file based on the position of the
+// app in the pod manifest.
+func EnvFilePath(root string, index int) string {
+	return filepath.Join(common.Stage1RootfsPath(root), RelEnvFilePath(index))
 }
 
-// ServiceWantPath returns the systemd default.target want symlink path for the
-// given imageID
-func ServiceWantPath(root string, imageID types.Hash) string {
-	return filepath.Join(common.Stage1RootfsPath(root), defaultWantsDir, ServiceUnitName(imageID))
+// ServiceWantPath returns the systemd default.target want symlink path based on the
+// position of the app in the pod manifest.
+func ServiceWantPath(root string, index int) string {
+	return filepath.Join(common.Stage1RootfsPath(root), defaultWantsDir, ServiceUnitName(index))
 }
 
 // InstantiatedPrepareAppUnitName returns the systemd service unit name for prepare-app
-// instantiated for the given root
-func InstantiatedPrepareAppUnitName(imageID types.Hash) string {
+// instantiated for the given root based on its position in the pod manifest.
+func InstantiatedPrepareAppUnitName(index int) string {
```

Comment:
`InstantiatedPrepareAppUnitName` reimplements the same thing as systemd-escape before the [function escape() in go-systemd](https://github.com/coreos/rkt/blob/master/Godeps/_workspace/src/github.com/coreos/go-systemd/unit/escape.go) existed, but it is a simplified version because we know the constraints on `imageID types.Hash` (or `index int` with this PR).

Now that go-systemd has a good and generic `UnitNamePathEscape` function, we could just use that instead of calling `strings.Replace()`. This could be done in a follow-up PR if you prefer, since it's not directly related to your change.

But if you go for the appnames as suggested by @jonboulle instead of the zero-indexed names, the escaping as currently implemented in `InstantiatedPrepareAppUnitName` is not good enough.